recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl

```diff
@@ -647,53 +668,71 @@ def main(argv):
+                        action='store_true')
     try:
         args = parser.parse_args()
     except IOError as e:
         print(e.strerror, file=sys.stderr)
-        exit(1)
+        return ErrorCode.INVALIDARG.value
 
-    update_input = input(
+    update_input = the_input(args.quiet,
         "\nWarning: Update may render device unbootable. Continue (y/N)? ")
     if not update_input == "y":
-        sys.exit(1)
+        return ErrorCode.CANCELED.value
 
     if args.force or args.reset:
-        erase_env_input = input("\nWarning: All U-Boot environment variables will be reset to factory settings. Continue (y/N)? ")
+        erase_env_input = the_input(args.quiet, "\nWarning: All U-Boot environment variables will be reset to factory settings. Continue (y/N)? ")
         if not erase_env_input == "y":
-            sys.exit(1)
+            return ErrorCode.CANCELED.value
     try:
         if args.force:
             # When forced, ignore all the checking and call the underlying updater directly
-            updater = FirmwareUpdate(args.firmware)
+            updater = FirmwareUpdate(args.firmware, args.bkup_dir)
         elif args.rollback:
-            updater = RollbackFirmware(args.firmware)
+            updater = RollbackFirmware(args.firmware, args.bkup_dir)
         else:
-            updater = ManagedFirmwareUpdate(args.firmware)
+            updater = ManagedFirmwareUpdate(args.firmware, args.bkup_dir)
 
         if args.rollback:
             rollback_confirm = input("\nRollback the firmware to the version before the upgrade (y/N)?")
             if not rollback_confirm == "y":
-                sys.exit(1)
-            updater.rollback_firmware()
-            sys.exit(1)
+                return ErrorCode.CANCELED.value
+            ret = updater.rollback_firmware()
+            if ErrorCode.ROLLBACKFAILED.value == ret:
+                return ret
+            return ErrorCode.SUCCESS.value
 
         envlist = None
         if not args.force and not args.reset:
             envlist = updater.get_preserved_uboot_env(args.preserve_list)
 
         if not args.no_backup:
-            updater.backup_firmware()
+            ret = updater.backup_firmware()
+            if ErrorCode.BKUPFAILED.value == ret:
+                return ret
         updater.update_firmware()
 
         if envlist:
             print("\n")
             for env in envlist:
                 print(env)
-            preserved_env_input = input(
+            preserved_env_input = the_input(args.quiet,
                 "\nWarning: Preserve the above env variables. Confirm(Y/n)? ")
             if not preserved_env_input == "n":
                 updater.update_uboot_env(envlist)
+
+        reboot_input = the_input(args.quiet,
+            "\nWarning: Completed. Please reboot the device. Reboot now(y/N)? ")
+        if not reboot_input == "y":
+            return ErrorCode.SUCCESS.value
+        os.system('reboot')
```

Comment:
If I'm not incorrect, this will reboot the device after FW update is finished. Could we please disable this for quiet mode? It's better to not reboot the device automatically without user's recognition.

Reply:
Yes, this script will reboot the device automatically with `quiet` mode. Hi @jan-kiszka @BaochengSu @AsuraZeng  any comments here?

Reply:
The `quiet mode` is that everything is selected by default, reboot is also the selected behavior.  So from my perspective, reboot with the quiet mode is ok.

Reply:
Quiet mode is mainly used in script, in that case, normally user also want to scripting-ly control the system behavior, so sounds to me, a non-reboot should be the default behavior under quiet mode, so that user can control the system reboot on their demands.

Also, before the reboot, user may want to collect the upgrading logs for further analysis, and also may want to use USER_LED to indicates the upgrading result, since in this case there is no way (FTDI, DP) to indicates the result other than LED.

So a default no-reboot sounds more reasonable to me.

But user needs to be aware of the risks of not immediately reboot, if some u-boot env manipulation happens before reboot, the firmware may corrupt.

Another way is to keep the reboot as the default quiet action, but also add a `no-reboot` parameter, but this would also introduce to handling the special case of combining the `no-reboot` with non-quiet mode upgrading, sounds not a elegant way.

Reply:
> If I'm not incorrect, this will reboot the device after FW update is finished. Could we please disable this for quiet mode? It's better to not reboot the device automatically without user's recognition.

+1

Reply:
@colpane @BaochengSu @AsuraZeng @karadaganil  Is there any more comments here?

Reply:
Hi @huaqianli,

Here is my understanding;
   if  quiet mode is enabled, all questions(excluded reboot) are "Yes".

From my point of view, "rollback_confirm" should be "No" since "-r|--rollback" is provided by the tool + platform builder cannot know the actual operation state since rollback_failed error code is just sent once no valid backup file found*. rollback seems perform like fimware update. So, the error code is not raised by update_firmware(included flash_operate()) method.


* https://github.com/huaqianli/meta-iot2050/blob/master/recipes-app/iot2050-firmware-update/files/iot2050-firmware-update.tmpl#L601

Reply:
I strongly suggest using of other cli arguments to separate rollback and reboot performing configs from "quiet" mode.

--auto-reboot      to specify firmware auto reboot configuration 
--auto-rollback   to specify firmware auto rollback configuration

Reply:
@huaqianli Please, comment on @karadaganil's comment. It is needed for us.

Reply:
What about we don't add `--auto-reboot` now and let the quiet mode not reboot the device?

My understanding is that auto reboot does not bring too much value to the quiet mode. Even if people really use auto-reboot, then a simple combination of `iot2050-fw-update --options && reboot` will do the trick.

Reply:
Regarding the rollback in quiet mode, what about to add a new return code, for example, named with `Rollback_Happened`, with non-zero return code, to indicates updating failed and rollback happened?

Or do you rather want to use the `-q` with no auto rollback and do the rollback in your own script / manually depends on your use case?

I am 50 50.